LibraBFT/Concrete/Properties/PreferredRound.agda

```diff
@@ -24,24 +25,216 @@ open import LibraBFT.Yasm.Yasm ℓ-RoundManager ℓ-VSFP ConcSysParms PeerCanSig
+   -- the grandparent of the block that v votes for (i.e., the preferred round rule)
+   → Σ (VoteParentData v'abs)
+           (λ vp → Cand-3-chain-head-round c2 ≤ Abs.round (vpParent vp))
+     ⊎ (VoteForRound∈ pk (v' ^∙ vRound) (v' ^∙ vEpoch) (v' ^∙ vProposedId) (msgPool pre))
+
+
+ -- Similarly in case the same step sends both v and v'
+ PR-ImplObligation₂ : Set (ℓ+1 ℓ-RoundManager)
+ PR-ImplObligation₂ =
+   ∀{pid s' outs pk}{pre : SystemState}
+   → (r  : ReachableSystemState pre)
+   -- For any honest call to /handle/ or /init/,
+   → (sps : StepPeerState pid (msgPool pre) (initialised pre) (peerStates pre pid) (s' , outs))
+   → ∀{mbr v vabs m v' v'abs m'}
+   → Meta-Honest-PK pk
+   -- For every vote v represented in a message output by the call
+   → v  ⊂Msg m  → send m ∈ outs
+   → (sig : WithVerSig pk v) → ¬ (∈GenInfo (ver-signature sig))
+   -- If v is really new and valid
+   → ¬ (MsgWithSig∈ pk (ver-signature sig) (msgPool pre)) -- ∄[ v'' ] VoteForRound∈ ... ?
+   → PeerCanSignForPK (StepPeer-post {pre = pre} (step-honest sps)) v pid pk
+   -- And if there exists another v' that is also new and valid
+   → v' ⊂Msg m'  → send m' ∈ outs
+   → (sig' : WithVerSig pk v') → ¬ (∈GenInfo (ver-signature sig'))
+   → ¬ (MsgWithSig∈ pk (ver-signature sig') (msgPool pre)) -- ∄[ v'' ] VoteForRound∈ ... ?
+   → PeerCanSignForPK (StepPeer-post {pre = pre} (step-honest sps)) v' pid pk
+   -- If v and v' share the same epoch and round
+   → v ^∙ vEpoch ≡ v' ^∙ vEpoch
+   → v ^∙ vRound < v' ^∙ vRound
+   → α-ValidVote 𝓔 v  mbr ≡ vabs
+   → α-ValidVote 𝓔 v' mbr ≡ v'abs
+   → (c2 : Cand-3-chain-vote vabs)
+   → Σ (VoteParentData v'abs)
+           (λ vp → Cand-3-chain-head-round c2 ≤ Abs.round (vpParent vp))
+
+  -- Next, we prove that given the necessary obligations,
+ module PR-Proof
    (sps-corr : StepPeerState-AllValidParts)
-   (Impl-PR1 : ImplObligation₁)
+   (Impl-IRO : VO.IncreasingRoundObligation)
+   (Impl-PR1 : PR-ImplObligation₁)
+   (Impl-PR2 : PR-ImplObligation₂)
    where
   -- Any reachable state satisfies the PR rule for any epoch in the system.
-  module _ (st : SystemState)(r : ReachableSystemState st)(𝓔 : EpochConfig) where
+  module _ (st : SystemState)(r : ReachableSystemState st) where
    -- Bring in 'unwind', 'ext-unforgeability' and friends
    open Structural sps-corr
    -- Bring in intSystemState
    open        PerState st r
    open        PerEpoch 𝓔
-   open import LibraBFT.Concrete.Obligations.PreferredRound 𝓔 (ConcreteVoteEvidence 𝓔) as PR
 
-   postulate  -- TODO-3: prove it
-     prr : PR.Type intSystemState
+
+   msgSentB4⇒VoteRound∈ : ∀ {v pk pool}
+                        → (vv : WithVerSig pk v)
+                        → (m : MsgWithSig∈ pk (ver-signature vv) pool)
+                        → VoteForRound∈ pk (v ^∙ vRound) (v ^∙ vEpoch) (v ^∙ vProposedId) pool
+   msgSentB4⇒VoteRound∈ {v} vv m
+      with sameSig⇒sameVoteData (msgSigned m) vv (msgSameSig m)
```

Comment:
You could simplify this by using `sameSig⇒sameVoteDataNoCol`, but it's no big deal because this proof is simple anyway, and hopefully one day we'll eliminate `meta-sha256-cr`, which will mean revisiting these kinds of proofs.

Reply:
You are right, I just did it!